[Track v1.4] New PR branch to serve as submodule for scikit-tree

adam2392 opened the pull request:
#### Reference Issues/PRs

As of v0.2 for sktree, we have decided we do not need a custom built and released via pypi scikit-learn fork. Instead, we just have to keep an updated fork branch here that maintains the changes under `tree/` and `ensemble/`.

This branch has significantly lower diff and less complexity compared to e.g. #44 


#### What does this implement/fix? Explain your changes.


#### Any other comments?


<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

adam2392 commented on sklearn/tree/_tree.pyx:
```diff
@@ -153,15 +169,91 @@ cdef struct StackRecord:
+        paths = tree.decision_path(X)
+        cdef int PARENT
+        cdef int CHILD
+        cdef int i
+        false_roots = {}
+        X_copy = {}
+        y_copy = {}
+        for i in range(X.shape[0]):
+            depth_i = paths[i].indices.shape[0] - 1
+            PARENT = depth_i - 1
+            CHILD = depth_i
+
+            if PARENT < 0:
+                parent_i = 0
+            else:
+                parent_i = paths[i].indices[PARENT]
+            child_i = paths[i].indices[CHILD]
+            left = 0
+            if tree.children_left[parent_i] == child_i:
+                left = 1
+
+            if (parent_i, left) in false_roots:
+                false_roots[(parent_i, left)][0] += 1
+                X_copy[(parent_i, left)].append(X[i])
+                y_copy[(parent_i, left)].append(y[i])
+            else:
+                false_roots[(parent_i, left)] = [1, depth_i]
+                X_copy[(parent_i, left)] = [X[i]]
+                y_copy[(parent_i, left)] = [y[i]]
+
+        X_list = []
+        y_list = []
+        for key, value in reversed(sorted(X_copy.items())):
+            X_list = X_list + value
+            y_list = y_list + y_copy[key]
+        cdef object X_new = np.array(X_list)
+        cdef cnp.ndarray y_new = np.array(y_list)
+
+        cdef Splitter splitter = self.splitter
+        splitter.init(X_new, y_new, sample_weight, missing_values_in_feature_mask)
```
@PSSF23 do you mind elaborating on what these LOC do? Perhaps you can submit another PR documenting these lines? 

I wonder if the error comes from these not setting certain variables correctly, so when build is called, something illegal happens

PSSF23 replied:
Sure

PSSF23 commented:
The failed [test](https://github.com/scikit-learn/scikit-learn/blob/a9611d09e8f7e35d40e658359519b698425ccd7f/sklearn/utils/estimator_checks.py#L1902) seems to have a very large number of classes, so I'll start searching around there.

PSSF23 reviewed (commented):
I cannot reproduce the error on my local machine with following code:

```python
from sklearn.ensemble import RandomForestClassifier

import numpy as np
from inspect import signature

rnd = np.random.RandomState(0)
n_samples = 30
X = rnd.uniform(size=(n_samples, 3))
y = np.arange(n_samples)

clf_1 = RandomForestClassifier()
clf_1.set_params(random_state=0)

func = getattr(clf_1, "fit", None)
func(X,y)
args = [p.name for p in signature(func).parameters.values()]

func = getattr(clf_1, "score", None)
func(X,y)
args = [p.name for p in signature(func).parameters.values()]

func = getattr(clf_1, "partial_fit", None)
func(X,y)
args = [p.name for p in signature(func).parameters.values()]
```

The code should replicate most of the `check_fit_score_takes_y` [test](https://github.com/scikit-learn/scikit-learn/blob/a9611d09e8f7e35d40e658359519b698425ccd7f/sklearn/utils/estimator_checks.py#L1902), but it runs smoothly every time. I also don't understand why only these 2 CIs failed when all should have the same test library.

- Linux_Runs pylatest_conda_forge_mkl 
- macOS pylatest_conda_mkl_no_openmp

adam2392 commented:
> I cannot reproduce the error on my local machine with following code:
> 
> ```python
> from sklearn.ensemble import RandomForestClassifier
> 
> import numpy as np
> from inspect import signature
> 
> rnd = np.random.RandomState(0)
> n_samples = 30
> X = rnd.uniform(size=(n_samples, 3))
> y = np.arange(n_samples)
> 
> clf_1 = RandomForestClassifier()
> clf_1.set_params(random_state=0)
> 
> func = getattr(clf_1, "fit", None)
> func(X,y)
> args = [p.name for p in signature(func).parameters.values()]
> 
> func = getattr(clf_1, "score", None)
> func(X,y)
> args = [p.name for p in signature(func).parameters.values()]
> 
> func = getattr(clf_1, "partial_fit", None)
> func(X,y)
> args = [p.name for p in signature(func).parameters.values()]
> ```
> 
> The code should replicate most of the `check_fit_score_takes_y` [test](https://github.com/scikit-learn/scikit-learn/blob/a9611d09e8f7e35d40e658359519b698425ccd7f/sklearn/utils/estimator_checks.py#L1902), but it runs smoothly every time. I also don't understand why only these 2 CIs failed when all should have the same test library.
> 
> * Linux_Runs pylatest_conda_forge_mkl
> * macOS pylatest_conda_mkl_no_openmp

If you try running the `HonestForestClassifier` instead, then the error seems to be able to be produced on my computer

PSSF23 commented:
@adam2392 Is `sktree` updated with the current submodule? After the resizing bug is fixed?

adam2392 commented:
Yeah it should be. The test is commented out tho rn.

PSSF23 commented:
@adam2392 I tried `from sktree import HonestForestClassifier` but still could not replicate the error. What python version are you using? I'm using 3.9. Also, can you check `sktree/_lib/sklearn/tree/_tree.pyx` line 282-292 to make sure the change is reflected there?

I doubt that this might be a python or conda compatibility issue.

PSSF23 reviewed (approved):
All CIs passed.

adam2392 committed:
Fix merge main

Signed-off-by: Adam Li <adam2392@gmail.com>

adam2392 committed:
Merge branch 'scikit-learn:main' into submodulev3

adam2392 committed:
Fix lint

Signed-off-by: Adam Li <adam2392@gmail.com>

adam2392 committed:
Merge main

Signed-off-by: Adam Li <adam2392@gmail.com>

adam2392 committed:
Merge branch 'scikit-learn:main' into submodulev3

adam2392 committed:
Merge branch 'scikit-learn:main' into submodulev3

adam2392 committed:
Merge branch 'scikit-learn:main' into submodulev3

adam2392 committed:
Merge branch 'scikit-learn:main' into submodulev3

adam2392 commented:
TODO: Change 

`int` -> `intp_t`
`double` -> `float64_t`
`SIZE_t` -> `intp_t`
`DTYPE_t` -> `float32_t`
`INT32_t` -> `int32_t`

see: https://github.com/scikit-learn/scikit-learn/pull/27352/files and related PRs

adam2392 committed:
Merge main

Signed-off-by: Adam Li <adam2392@gmail.com>

adam2392 commented:
> TODO: Change
> 
> `int` -> `intp_t` `double` -> `float64_t` `SIZE_t` -> `intp_t` `DTYPE_t` -> `float32_t` `INT32_t` -> `int32_t`
> 
> see: https://github.com/scikit-learn/scikit-learn/pull/27352/files and related PRs

This was accomplished in https://github.com/neurodata/scikit-learn/pull/53/commits/9a5d91b5827ef054599cf7a515421fc27d88e1fb

adam2392 committed:
Merge branch 'scikit-learn:main' into submodulev3

adam2392 committed:
Fix utils.pyx typing

Signed-off-by: Adam Li <adam2392@gmail.com>

adam2392 committed:
Try absolute import

Signed-off-by: Adam Li <adam2392@gmail.com>

adam2392 committed:
Try again

Signed-off-by: Adam Li <adam2392@gmail.com>

adam2392 committed:
Update import path

Signed-off-by: Adam Li <adam2392@gmail.com>

adam2392 committed:
Make submodule install easier

Signed-off-by: Adam Li <adam2392@gmail.com>

adam2392 committed:
Change ctypedef in random

Signed-off-by: Adam Li <adam2392@gmail.com>

adam2392 committed:
Revert UINT32_t

Signed-off-by: Adam Li <adam2392@gmail.com>

adam2392 committed:
Merge branch 'scikit-learn:main' into submodulev3

adam2392 committed:
Change cnp.float64 to float64_t

Signed-off-by: Adam Li <adam2392@gmail.com>

adam2392 committed:
Merging main

Signed-off-by: Adam Li <adam2392@gmail.com>

adam2392 committed:
Merge branch 'scikit-learn:main' into submodulev3

adam2392 committed:
Merge branch 'scikit-learn:main' into submodulev3

adam2392 committed:
Merge branch 'scikit-learn:main' into submodulev3

adam2392 committed:
Merge main

Signed-off-by: Adam Li <adam2392@gmail.com>

adam2392 committed:
Merge branch 'scikit-learn:main' into submodulev3

adam2392 committed:
Merge branch 'scikit-learn:main' into submodulev3

adam2392 committed:
Remove base estimator deprecated name

Signed-off-by: Adam Li <adam2392@gmail.com>

adam2392 committed:
Merge branch 'submodulev3' of https://github.com/neurodata/scikit-learn into submodulev3

adam2392 committed:
Merge branch 'main' into submodulev3

adam2392 committed:
Merge branch 'main' into submodulev3

adam2392 committed:
Merge branch 'main' into submodulev3

adam2392 committed:
Merge branch 'main' into submodulev3

adam2392 committed:
Merging main

Signed-off-by: Adam Li <adam2392@gmail.com>

adam2392 committed:
Merge branch 'main' into submodulev3

adam2392 committed:
Merge branch 'main' into submodulev3

adam2392 committed:
Make sure build_tree returns self

Signed-off-by: Adam Li <adam2392@gmail.com>

adam2392 committed:
Merge branch 'main' into submodulev3

adam2392 committed:
Allow max samples to be higher

Signed-off-by: Adam Li <adam2392@gmail.com>

adam2392 committed:
Factor out construct trees API

Signed-off-by: Adam Li <adam2392@gmail.com>

adam2392 committed:
Merge branch 'main' into submodulev3

adam2392 committed:
Merge branch 'main' into submodulev3

adam2392 committed:
Allow extra args in cinit

Signed-off-by: Adam Li <adam2392@gmail.com>

adam2392 committed:
Merge branch 'main' into submodulev3

adam2392 committed:
Merge branch 'main' into submodulev3

adam2392 committed:
Migrate n_constant_features within SplitRecord

Signed-off-by: Adam Li <adam2392@gmail.com>

adam2392 commented:
e1224a483 introduces a major change, where `n_constant_features` is migrated to WITHIN the `SplitRecord`. This is to enable over-riding what gets passed from parent node to child node.

I.e. MultiviewSplitRecord could in principle override SplitRecord and store additional `n_constant_features`. One per feature set. Ofc, I don't think this is actually the best design cuz the idea of "multi-view" really breaks away from the underlying assumption that there is only one feature set.

adam2392 committed:
Export shift_missing_values_to_left_if_required

Signed-off-by: Adam Li <adam2392@gmail.com>

adam2392 committed:
demo pr

Signed-off-by: Adam Li <adam2392@gmail.com>

adam2392 committed:
Demo

Signed-off-by: Adam Li <adam2392@gmail.com>

adam2392 committed:
Merge branch 'main' into nconstantv2

adam2392 committed:
Benchmarks

Signed-off-by: Adam Li <adam2392@gmail.com>

adam2392 committed:
Merge main

Signed-off-by: Adam Li <adam2392@gmail.com>

adam2392 committed:
Merge main

Signed-off-by: Adam Li <adam2392@gmail.com>

adam2392 committed:
Bench size

Signed-off-by: Adam Li <adam2392@gmail.com>

adam2392 committed:
Adding parentrecord

Signed-off-by: Adam Li <adam2392@gmail.com>

adam2392 committed:
Merge branch 'nconstantv2' of https://github.com/adam2392/scikit-learn into nconstantv2

adam2392 committed:
Fix lint

Signed-off-by: Adam Li <adam2392@gmail.com>

adam2392 committed:
Fix bestfirst

Signed-off-by: Adam Li <adam2392@gmail.com>

adam2392 committed:
Init parent record

Signed-off-by: Adam Li <adam2392@gmail.com>

adam2392 committed:
Merge branch 'main' into nconstantv2

adam2392 committed:
Merge branch 'main' into nconstantv2

adam2392 committed:
Merge branch 'main' into submodulev3

adam2392 committed:
Merge branch 'main' into nconstantv2

adam2392 committed:
Address thomas comments' -s

Signed-off-by: Adam Li <adam2392@gmail.com>

adam2392 committed:
Merge branch 'nconstantv2' of https://github.com/adam2392/scikit-learn into nconstantv2

adam2392 committed:
Merge branch 'main' into submodulev3

adam2392 committed:
Revert to unit32_t

Signed-off-by: Adam Li <adam2392@gmail.com>

adam2392 committed:
Fix build

Signed-off-by: Adam Li <adam2392@gmail.com>

adam2392 committed:
UPdate submodule commit and remove extraneous code

Signed-off-by: Adam Li <adam2392@gmail.com>

adam2392 committed:
Fixing LOC

Signed-off-by: Adam Li <adam2392@gmail.com>

adam2392 committed:
Fix update tree node

Signed-off-by: Adam Li <adam2392@gmail.com>

adam2392 committed:
Fix ci

Signed-off-by: Adam Li <adam2392@gmail.com>

adam2392 committed:
Fix ci

Signed-off-by: Adam Li <adam2392@gmail.com>

adam2392 committed:
MAINT Fix builder partial fit (#62)

#### Reference Issues/PRs
Fixes state of builder_ to not need to be maintained.

Prolly needs unit-tests to determine if this "functions as desired".
I.e.
- changing datatype of X over multiple partial fits should fail nicely, 
- changing datatype of y
- classification and regression

#### What does this implement/fix? Explain your changes.


#### Any other comments?


<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:

https://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

---------

Signed-off-by: Adam Li <adam2392@gmail.com>

adam2392 committed:
Try again for partial fit

Signed-off-by: Adam Li <adam2392@gmail.com>

adam2392 committed:
Merging main

Signed-off-by: Adam Li <adam2392@gmail.com>

adam2392 committed:
Adding inline comment

Signed-off-by: Adam Li <adam2392@gmail.com>

adam2392 committed:
Merge branch 'main' into submodulev3

adam2392 committed:
Merging main

Signed-off-by: Adam Li <adam2392@gmail.com>

adam2392 committed:
Merging main

Signed-off-by: Adam Li <adam2392@gmail.com>

adam2392 committed:
Merging main

Signed-off-by: Adam Li <adam2392@gmail.com>

adam2392 committed:
Simplify cython partition api

Signed-off-by: Adam Li <adam2392@gmail.com>

adam2392 committed:
Simplify cython partition api

Signed-off-by: Adam Li <adam2392@gmail.com>

adam2392 committed:
cimport _build pruned tree

Signed-off-by: Adam Li <adam2392@gmail.com>

adam2392 committed:
Merge branch 'main' into submodulev3

adam2392 committed:
Merge branch 'main' into submodulev3

adam2392 committed:
Merging main

Signed-off-by: Adam Li <adam2392@gmail.com>

adam2392 committed:
Make sure its defined

Signed-off-by: Adam Li <adam2392@gmail.com>

adam2392 committed:
Merged latest changes from sklearn upstream 07/20/24

Signed-off-by: Adam Li <adam2392@gmail.com>

adam2392 committed:
Merging main 080724

Signed-off-by: Adam Li <adam2392@gmail.com>

adam2392 committed:
Update wrt main

Signed-off-by: Adam Li <adam2392@gmail.com>

adam2392 committed:
Merging main

Signed-off-by: Adam Li <adam2392@gmail.com>

adam2392 committed:
Merging main

Signed-off-by: Adam Li <adam2392@gmail.com>

adam2392 committed:
Merging main

Signed-off-by: Adam Li <adam2392@gmail.com>

adam2392 committed:
Merging main

Signed-off-by: Adam Li <adam2392@gmail.com>

adam2392 committed:
Reverting

Signed-off-by: Adam Li <adam2392@gmail.com>

adam2392 commented:
We should remove binning for now as it is a untested feature...

adam2392 committed:
Fixed

Signed-off-by: Adam Li <adam2392@gmail.com>

adam2392 committed:
Fixed

Signed-off-by: Adam Li <adam2392@gmail.com>

adam2392 committed:
Fixed

Signed-off-by: Adam Li <adam2392@gmail.com>

adam2392 committed:
Fixed

Signed-off-by: Adam Li <adam2392@gmail.com>

adam2392 committed:
Reverting back to imports

Signed-off-by: Adam Li <adam2392@gmail.com>

adam2392 committed:
Fixed

Signed-off-by: Adam Li <adam2392@gmail.com>

adam2392 committed:
Fix validate

Signed-off-by: Adam Li <adam2392@gmail.com>

adam2392 committed:
Fix partial fit

Signed-off-by: Adam Li <adam2392@gmail.com>

adam2392 committed:
Fix partial fit

Signed-off-by: Adam Li <adam2392@gmail.com>

adam2392 committed:
Merge branch 'main' into submodulev3

adam2392 committed:
Adding unit test for partial fit

Signed-off-by: Adam Li <adam2392@gmail.com>

adam2392 committed:
Merge branch 'main' into submodulev3

adam2392 committed:
Merging main

Signed-off-by: Adam Li <adam2392@gmail.com>

adam2392 committed:
Merge main

Signed-off-by: Adam Li <adam2392@gmail.com>

PSSF23 committed:
FIX remove regressor multi_label tag (#71)

<!--
Thanks for contributing a pull request! Please ensure you have taken a
look at
the contribution guidelines:
https://github.com/scikit-learn/scikit-learn/blob/main/CONTRIBUTING.md
-->

#### Reference Issues/PRs
https://github.com/neurodata/treeple/pull/339
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull
requests
you resolved, so that they will automatically be closed when your pull
request
is merged. See
https://github.com/blog/1506-closing-issues-via-pull-requests
-->


#### What does this implement/fix? Explain your changes.


#### Any other comments?


<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:

https://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

PSSF23 committed:
FIX remove xfail_checks (#74)

According to errors in https://github.com/neurodata/treeple/pull/361,
`xfail_checks` is not available in the tags.

PSSF23 committed:
FIX add argument to _check_sample_weight (#73)

Fix the following bug:
```
>>> from treeple.ensemble import ObliqueRandomForestClassifier
>>> from sklearn.datasets import make_classification
>>> X,y = make_classification(n_samples=1000, n_features=4,n_informative=2,n_redundant=0,random_state=0,shuffle=False)
>>> clf = ObliqueRandomForestClassifier(max_depth=2, random_state=0)
>>> clf.fit(X,y)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/pssf23/miniconda3/envs/temp/lib/python3.10/site-packages/sklearn/base.py", line 1363, in wrapper
    return fit_method(estimator, *args, **kwargs)
  File "/Users/pssf23/miniconda3/envs/temp/lib/python3.10/site-packages/treeple/_lib/sklearn/ensemble/_forest.py", line 597, in fit
    self._construct_trees(
  File "/Users/pssf23/miniconda3/envs/temp/lib/python3.10/site-packages/treeple/_lib/sklearn/ensemble/_forest.py", line 663, in _construct_trees
    trees = Parallel(
  File "/Users/pssf23/miniconda3/envs/temp/lib/python3.10/site-packages/sklearn/utils/parallel.py", line 82, in __call__
    return super().__call__(iterable_with_config_and_warning_filters)
  File "/Users/pssf23/miniconda3/envs/temp/lib/python3.10/site-packages/joblib/parallel.py", line 1986, in __call__
    return output if self.return_generator else list(output)
  File "/Users/pssf23/miniconda3/envs/temp/lib/python3.10/site-packages/joblib/parallel.py", line 1914, in _get_sequential_output
    res = func(*args, **kwargs)
  File "/Users/pssf23/miniconda3/envs/temp/lib/python3.10/site-packages/sklearn/utils/parallel.py", line 147, in __call__
    return self.function(*args, **kwargs)
  File "/Users/pssf23/miniconda3/envs/temp/lib/python3.10/site-packages/treeple/_lib/sklearn/ensemble/_forest.py", line 211, in _parallel_build_trees
    tree._fit(
  File "/Users/pssf23/miniconda3/envs/temp/lib/python3.10/site-packages/treeple/_lib/sklearn/tree/_classes.py", line 395, in _fit
    sample_weight = _check_sample_weight(sample_weight, X, DOUBLE)
TypeError: _check_sample_weight() takes 2 positional arguments but 3 were given
```

PSSF23 committed:
FIX remove checks on X and y to allow 3D data (#75)

PSSF23 committed:
FIX correct shape assignment for 3D data (#76)
